pkg/plugin/v3/init.go

```diff
@@ -87,6 +88,8 @@ func (p *initPlugin) BindFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&p.license, "license", "apache2",
 		"license to use to boilerplate, may be one of 'apache2', 'none'")
 	fs.StringVar(&p.owner, "owner", "", "owner to add to the copyright")
+	fs.BoolVar(&p.componentConfig, "componentconfig", false,
+		"create a versioned componentconfig file, may be 'true' or 'false'")
```

Comment:
Hi @christopherhein, 

Why not. create a new plugin for it? IMO we could have a ComponentConfigPlugin such as the INIT plugin which means that if users would like to use this option at the same way that the use the addon as well. 

Also, why it is a breaking change? Is only because of the upper version of controller-runtime?

Reply:
I implemented this as not a breaking change other than the CR version which is going to be done in #1746, the original issue had mentioned to make componentconfig the default and I'm thinking it's the right idea to do that until we have in controller runtime automatic version conversion to at least `v1beta1`.

Reply:
Regarding doing this as a plugin; can you point me to where I should read about for what that would entail?

Reply:
`pkg/plugin` has the interfaces for init, edit, create API and create webhook plugins, and `pkg/plugin/vX` implement those interfaces. The calls to this interfaces is done from `pkg/cli` from the file with the corresponding names.

- I think Camila means creating a new type of plugin (requiring new interfaces similar to `pkg/plugin.Init` and `pkg/plugin.InitSubcommand`) but I'm not sure if it is a bit of an over kill. I think it may be a good thing to add to the edit plugin, which right now is only used to turn on multi-group.
- We could have a `edit component-config` command that enables or disables this behavior but in this case we would only be targeting `v3` as it would require a new field in the project configuration file which would be a breaking change with `v2`. 
- We could also make the `edit component-config` scaffold the relevant files instead of enabling the feature, which would be `v2`-compatible.

Reply:
@Adirio and @camilamacedo86 I'd considered doing it like the edit command first the difficult thing is we'd have to rewrite a lot of the init code to support "edit component-config" which isn't currently addressable with markers. 

For example removing all the flag declarations in `main.go` so that componentconfig takes precedence  https://github.com/kubernetes-sigs/kubebuilder/pull/1790/files#diff-6d104501fc7dc73bb67118a31ea122a23d55af27d443c7b96029054238978a36R113-R124  or where I have to remove all the flags from the config/manager and config/default/auth_proxy - https://github.com/kubernetes-sigs/kubebuilder/pull/1790/files#diff-d524621fbcb369b36df89df9480d29cb86e7779e7b6b2dec466f66f11742cae6L46-R61

I'd like to keep this from being a breaking v2 change as much as possible. There is a longer-term implementation that once we switch to it being the default we can make it so that you can supply a custom type and it would scaffold that for your controller see https://github.com/kubernetes-sigs/controller-runtime/blob/master/examples/configfile/custom/main.go#L49-L53 which might be a good place to use the `edit` changes.

Reply:
Hi @christopherhein, 

I catcher its end goal now. Would really be a common bahaviour users want to init a project with config component by default? Would not possible they want to change for this option after the project be already init? 

So, WDYT
- we add new attribute into the PROJECT file to let us know if the project has or not this option such as the multigroup one. (IMO: we need to store all inputs otherwise will be impossible one day we achieve end goals such as have a migration tool for example) 
- then, could we not notify and re-scaffold the main and auth_proxy if we kb edit configcomponent=true/false`? Following an example:

> kubebuilder edit --component-config
> Beware that this command will re-generate the files: main.go and config/default/manager_auth_proxy_patch.yaml. Would you like to continue? [y/n]

Reply:
I am trying to figure out a way to allow users to have this option as they have the multigroup support one or addon. Since it is an option that shows that add to the init might not really be useful in daily bases.  However, feel free to raise other ideas as well. 

Also, we might need to have a new testdata/ for v3 with this scenario as well.  See: https://github.com/kubernetes-sigs/kubebuilder/tree/master/testdata/ and the projects are build from https://github.com/kubernetes-sigs/kubebuilder/blob/master/generate_testdata.sh.

Reply:
That seems like a reasonable way of handling it with `edit`. Asking people before it regenerates could be nice. Are we doing that with multi-group to support moving controllers to `apis/` from `api/`?

@vincepri since you are interested in using this for `v1alpha4` CAPI does this experience raise any issues?

Reply:
`kubebuilder edit --multigroup` doesn't regenerate the projects, you are requested to do that manually. It just sets a flag in the project configuration file that will make future calls to `kubebuilder create ...` use the corresponding directory structure.

Reply:
Hmm, so if I'm reading what you're saying @Adirio correctly it doesn't sound like this would be the right candidate for `edit`, since we'd need to make subtle changes like adding `.AndFromOrDie(ctrl.ConfigFile())` to the `ctrl.Options{}` initialization which wouldn't work to add Markers for. All that being keeping it as a flag on `init` seems to be the best path. 

Am and I understanding properly?

Reply:
Could that be solved by creating a options first, placing a marker for your function call after, and then creating the manager below that? We could probably also have better error handling than `AndFromOrDie` too.

Reply:
H @christopherhein and @Adirio,

Actually, the edit multigroup does: 
- Update the dockerfile accordingly
- Update the project file as well
- All commands used the value in the Project file to generate the manifests accordingly
- And then, we have a few tasks to improve this behaviour too. See : https://github.com/kubernetes-sigs/kubebuilder/issues/1640

For multigroup scenario, is not required to re-generated the files. However, for the case of componentconfig the easy path shows do that at least at the first moment do it. We can try to improve the behaviour after as well.